Feature: add application scopes framework and imp health scope

wonderflow opened the pull request:
This PR is mainly to add application scope framework and health scope

This PR lasts several weeks mostly because the spec is changing.

https://github.com/microsoft/hydra-spec/issues/130
https://github.com/microsoft/hydra-spec/issues/144
https://github.com/microsoft/hydra-spec/pull/133
https://github.com/microsoft/hydra-spec/pull/127 
https://github.com/microsoft/hydra-spec/pull/177
https://github.com/microsoft/hydra-spec/pull/196

## add application scopes framework 

- [x] 1. add framework for network/health
- [x] 2. create health scope instance
- [x] 3. check component scope overlap
- [x] 4. add/remove Component to scope

## add health scope

- [x] 1. create a sub project to implement healthscope
- [x] 2. periodically aggregate status of health scope
- [x] 3. create a service to make health scope available for others
- [x] 4. docs for how to use it.


This is same with #160 as we have changed repo, I have to make a new PR.

fixes #9

wonderflow commented:
It seems the version of openssl installed by base image is 1.1.0 which didn't match the requirement of the updated lib.

I'm trying to figure out whether I should use a new base image

wonderflow commented:
@technosophos @hongchaodeng  I think this PR is ready for review now.

I have reset all the commits and split them into four big commits. Each commit is a different part of code so you can review them one by one.

wonderflow commented on Dockerfile:
```diff
@@ -1,27 +1,86 @@
+		| cut -d: -f1 \
+		| sort -u \
+		| xargs -r apt-mark manual \
+&& apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false
```
kube-rs 0.6.1 need openssl 1.1.1 while using apt-get in debian:stretch-slim didn't work, so I have to compile from source here.

hongchaodeng replied:
Maybe add the comments in the Dockerfile?

hongchaodeng commented on src/schematic/scopes/network.rs:
```diff
@@ -0,0 +1,117 @@
+use crate::schematic::configuration::ComponentConfiguration;
+/// Network scope is defined as https://github.com/oam-dev/spec/blob/master/4.application_scopes.md#network-scope
+/// Now we don't really implement network scope, this is just a framework as the spec describe.
```
add _TODO_

hongchaodeng commented on healthscope/README.md:
```diff
@@ -0,0 +1,197 @@
+# Health Scope Controller
+
+Health Scope Controller used for periodically check health scope crd and check the health of all the components related.
```
```suggestion
HealthScope controller is used to periodically check the health of all the components and update health scope custom resources.
```

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+                        };
+                    }
+                }
+                Err(e) => log::error!("get health scope list err {}", e),
```
```suggestion
                Err(e) => error!("get health scope list err {}", e),
```

hongchaodeng commented on healthscope/README.md:
```diff
@@ -0,0 +1,197 @@
+
+## What will health scope controller do?
+
+1. periodically check all component health and update the CR status.
```
```suggestion
1. periodically check health status of components and update the HealthScope resource status.
```

hongchaodeng commented on healthscope/README.md:
```diff
@@ -0,0 +1,197 @@
+
+## How to install?
+
+You could use helm to install it:
```
```suggestion
Use helm to install:
```

hongchaodeng commented on healthscope/README.md:
````diff
@@ -0,0 +1,197 @@
+
+You could use helm to install it:
+
+```shell script
````
```suggestion
```shell
```

wonderflow replied:
my IDE auto complete `shell` to `shell script `

hongchaodeng commented on healthscope/README.md:
```diff
@@ -0,0 +1,197 @@
+
+## How to use?
+
+After healthscope was installed by helm charts, you could setup an endpoint to it.
```
```suggestion
After healthscope controller is installed by helm charts, setup an endpoint to it.
```

hongchaodeng commented on healthscope/README.md:
```diff
@@ -0,0 +1,197 @@
+## How to use?
+
+After healthscope was installed by helm charts, you could setup an endpoint to it.
+By default the charts will use ClusterPort, you could use port mapper to make health scope accessible.
```
```suggestion
By default the charts will use ClusterPort, we can use port-forward to access healthscope controller.
```

hongchaodeng commented on healthscope/README.md:
````diff
@@ -0,0 +1,197 @@
+kubectl port-forward $POD_NAME 8080:80
+```
+
+Then you'll be able to visit http://127.0.0.1:8080 to get health scope status.
````
```suggestion
Then we'll be able to visit http://127.0.0.1:8080 to query status of healthscope resource.
```

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+                }
+                Err(e) => log::error!("get health scope list err {}", e),
+            }
+            cnt = (cnt + 1) % 10;
```
remove debugging code?

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+                debug!("health scope aggregate loop running...");
+            }
+            //FIXME: we could change this to use an informer if we have a runtime controller queue
+            std::thread::sleep(std::time::Duration::from_secs(1));
```
1s as default might be too frequent. Maybe 5s?

wonderflow replied:
ok

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+    health_scope_watch.join().unwrap()
+}
+
+use std::{
```
Put this in the head of file?

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+}
+
+type BoxFut = Box<dyn Future<Item = Response<Body>, Error = hyper::Error> + Send>;
+fn serve_health(req: Request<Body>) -> BoxFut {
```
add some comments?

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+    Box::new(future::ok(response))
+}
+
+fn request_health(instance_name: String) -> Result<String, Error> {
```
add some comments?

hongchaodeng commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,363 @@
+    Ok(health.to_string())
+}
+
+fn aggregate_health(
```
aggregate_health -> aggregate_components_health

hongchaodeng commented:
ref https://github.com/oam-dev/spec/pull/156

wonderflow commented:
@hongchaodeng  thanks very much, all comments are fixed in the last commit

fibonacci1729 commented:
@wonderflow i'll review this today.

technosophos commented on healthscope/src/main.rs:
```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+// FIXME kube-rs client doesn't support async call so we have to wrap it in HealthFuture here.
```
This looks like a solid solution while we wait for the Kube upstream.

technosophos commented:
The rebase that needs to be done is really simple, so feel free to merge as soon as you have rebased. I am excited to see this new addition!

wonderflow commented:
Thanks @technosophos , I've rebased.

I noticed changes from https://github.com/oam-dev/rudr/pull/375 has been overrided by #366. I rebased it back here.

wonderflow commented:
CI has passed, merging
